### PR TITLE
Fixed issue with cache not retaining refresh token

### PR DESCRIPTION
**Pull request by stevehobbsdev**
### Description

This PR corrects a design problem with the new local storage cache (and the original in-memory cache) where the refresh token was not being retained.

Currently, when the `expiresIn` value is reached and the access token should be dropped/renewed, the cache is cleared. Unfortunately, this also clears the refresh token.

This PR corrects that by retaining the refresh token when the access token expires, by stripping everything from the cache data except the refresh token, which can then be used to fetch a new access token.

### Testing

Manual testing in the playground, as well a new unit tests to cover this scenario.

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [X] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [X] All active GitHub checks for tests, formatting, and security are passing
- [X] The correct base branch is being used, if not `master`

### CODE REVIEW

**Comment by damieng** [comment]
Looks like a binary .tgz is included by mistake?

File: cypress/integration/getTokenSilently.js [hunk 1]
```diff
@@ -55,7 +55,7 @@ describe('getTokenSilently', function() {
           cy.toggleSwitch('local-storage');
 
           cy.login().then(() => {
-            cy.reload();
+            cy.reload().wait(5000);
```
Comment (damieng) [inline cypress/integration/getTokenSilently.js]: Why is there a new 5 second delay here?

Comment (stevehobbsdev) [reply cypress/integration/getTokenSilently.js]: This test intermittently fails because the UI isn’t ready yet, I’m hoping the delay will fix that. Seems to in the tests I’ve run locally, will see how it pans out over time.

**Comment by stevehobbsdev** [comment]
Oops, will remove the .tgz

File: cypress/integration/getTokenSilently.js [hunk 2]
```diff
@@ -82,7 +82,8 @@ describe('getTokenSilently', function() {
             cy.toggleSwitch('use-cache');
 
             cy.login().then(() => {
-              cy.toggleSwitch('refresh-tokens').wait(250);
+              cy.toggleSwitch('refresh-tokens').wait(1000);
```
Comment (lbalmaceda) [inline cypress/integration/getTokenSilently.js]: is this also related to the 5 sec delay comment above?

File: src/cache.ts [hunk 3]
```diff
@@ -59,27 +64,39 @@ export class LocalStorageCache implements ICache {
     );
 
     setTimeout(() => {
-      window.localStorage.removeItem(cacheKey);
+      const payload = this.getPayload(cacheKey);
+
+      if (!payload || !payload.body) return;
```
Comment (lbalmaceda) [inline src/cache.ts]: is this the right style, or do we need to use braces?

File: __tests__/cache.test.ts [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import { InMemoryCache, LocalStorageCache } from '../src/cache';
+import { _ } from 'core-js';
```
[the remark this answers is not in the document]
Comment (stevehobbsdev) [reply __tests__/cache.test.ts]: Removed

File: __tests__/cache.test.ts [hunk 5]
```diff
@@ -54,17 +55,48 @@ describe('InMemoryCache', () => {
...
+          exp: new Date().getTime() / 1000 + 2
+        },
+        user: { name: 'Test' }
+      }
+    });
+
+    jest.advanceTimersByTime(799);
+    expect(Object.keys(cache.cache).length).toBe(1);
+
+    jest.advanceTimersByTime(1);
```
[the remark this answers is not in the document]
Comment (stevehobbsdev) [reply __tests__/cache.test.ts]: This is a preexisting test from before this work went in, but essentially it tests that the cache is in one state right up to the expiry time, then as soon as the time advances onto the expiry time the cache should be in another state (i.e. it's been removed by a `setTimeout`).

Here, 800ms is used because Luis had a calculation in that defines the expiry time as `Math.min(expiresIn, expTime) * 1000 * 0.8`, and the `expiresIn` for this test is defined as 1000ms.

File: __tests__/cache.test.ts [hunk 6]
```diff
@@ -127,59 +160,64 @@ describe('LocalStorageCache', () => {
...
-          exp: nowSeconds() + 100
-        }
-      }
+  describe('cache.get', () => {
+    it('can retrieve an item from the cache', () => {
+      localStorage.setItem(
+        '@@auth0spajs@@::__TEST_CLIENT_ID__::__TEST_AUDIENCE__::__TEST_SCOPE__',
+        JSON.stringify({
+          body: defaultEntry,
+          expiresAt: nowSeconds() + 86400
```
[the remark this answers is not in the document]
Comment (stevehobbsdev) [reply __tests__/cache.test.ts]: Done

File: __tests__/index.test.ts [hunk 7]
```diff
@@ -1270,6 +1270,16 @@ describe('Auth0', () => {
...
         });
+
+        it('continues method execution when there is a value from the cache but no access token', async () => {
+          const { auth0, utils, cache } = await setup();
+
+          cache.get.mockReturnValue({});
+
+          await auth0.getTokenSilently();
+
+          expect(utils.encode).toHaveBeenCalledWith(TEST_RANDOM_STRING);
```
[the remark this answers is not in the document]
Comment (stevehobbsdev) [reply __tests__/index.test.ts]: Done, copied a variant of it

File: cypress/integration/getTokenSilently.js (hunk 2 above)
Comment (stevehobbsdev) [reply cypress/integration/getTokenSilently.js]: Yes, seems to have cured an intermittant failure. At least locally..

File: src/cache.ts (hunk 3 above)
Comment (stevehobbsdev) [reply src/cache.ts]: You're right, I've now applied a style consistant with the rest of the app.

File: __tests__/cache.test.ts [hunk 8]
```diff
@@ -127,59 +160,64 @@ describe('LocalStorageCache', () => {
...
+        '@@auth0spajs@@::__TEST_CLIENT_ID__::__TEST_AUDIENCE__::__TEST_SCOPE__',
+        JSON.stringify({
+          body: {
+            client_id: '__TEST_CLIENT_ID__',
+            audience: '__TEST_AUDIENCE__',
+            scope: '__TEST_SCOPE__',
+            id_token: '__ID_TOKEN__',
+            access_token: '__ACCESS_TOKEN__',
+            refresh_token: '__REFRESH_TOKEN__',
+            expires_in: -10,
```
[the remark this answers is not in the document]
Comment (stevehobbsdev) [reply __tests__/cache.test.ts]: I've reworked these tests to use a mocked up `Date.now` instead of a negative value, so they should look better.

File: __tests__/cache.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply __tests__/cache.test.ts]: Do you think something of the above, simplified of course, can be put as a comment to illustrate this testing intention? So future _me_ follows it

[the remark this answers is not in the document]
Comment (stevehobbsdev) [reply __tests__/cache.test.ts]: Done!
